src/audio_unit/delay.rs

```diff
@@ -1,32 +1,102 @@
 use crate::{ring_buffer, AudioUnit, Result};
+use anyhow::anyhow;
 use cpal::StreamConfig;
 use ringbuf::{Consumer, Producer, RingBuffer};
+use wmidi::MidiMessage;
+
+pub type DelayMs = u32;
+
+type MidiMessageHandler = Box<dyn Fn(&[MidiMessage<'static>]) -> Option<DelayMs> + Send + Sync>;
 
 pub struct Delay {
+    delay_ms: DelayMs,
+    midi_message_handler: MidiMessageHandler,
+    stream_config: StreamConfig,
     producer: Producer<f32>,
     consumer: Consumer<f32>,
 }
 
 impl Delay {
-    pub fn new(stream_config: &StreamConfig, delay_ms: u32) -> Result<Self> {
-        let delay_num_frames = (delay_ms as f32 / 1_000.0) * stream_config.sample_rate.0 as f32;
-        let delay_num_samples = delay_num_frames as usize * stream_config.channels as usize;
+    pub const MIN_DELAY_MS: DelayMs = 0;
+    pub const MAX_DELAY_MS: DelayMs = 10000;
 
-        let ring = RingBuffer::new(delay_num_samples * 2);
+    pub fn new(
+        stream_config: &StreamConfig,
+        midi_message_handler: MidiMessageHandler,
+        delay_ms: DelayMs,
+    ) -> Result<Self> {
+        Self::validate_delay(delay_ms)?;
+
+        // size the ring buffer so that it can accomodate the largest allowed delay
+        let ring = RingBuffer::new(Self::delay_num_samples(stream_config, Self::MAX_DELAY_MS) * 2);
         let (mut producer, consumer) = ring.split();
 
-        ring_buffer::write_empty_samples(&mut producer, delay_num_samples)?;
+        ring_buffer::write_empty_samples(
+            &mut producer,
+            Self::delay_num_samples(stream_config, delay_ms),
+        )?;
+
+        Ok(Self {
+            delay_ms,
+            midi_message_handler,
+            stream_config: stream_config.clone(),
+            producer,
+            consumer,
+        })
+    }
+
+    fn validate_delay(delay_ms: DelayMs) -> Result<()> {
+        if delay_ms > Self::MAX_DELAY_MS {
+            return Err(anyhow!(
+                "Delay must be <= {}, but was {}",
+                Self::MAX_DELAY_MS,
+                delay_ms
+            ));
+        } else {
+            Ok(())
+        }
+    }
+
+    fn delay_num_samples(stream_config: &StreamConfig, delay_ms: u32) -> usize {
+        let delay_num_frames = (delay_ms as f32 / 1_000.0) * stream_config.sample_rate.0 as f32;
+        delay_num_frames as usize * stream_config.channels as usize
+    }
+
+    fn set_delay_ms(&mut self, delay_ms: u32) -> Result<()> {
```

Comment:
This function works in the sense the delay is correctly adjusted as MIDI messages come in, but it sounds awful in the transition period. It's not clear to me what the expected behavior here should be. Any ideas?

Reply:
I'd have to wrap my head more around what exactly this is doing + why it sounds bad during the transition period